### ISSUE TITLE
FPN-12 | Phase 3 | Integrate XGBoost and Autoencoder into API with Hybrid Decision

### ISSUE THREAD

**Issue by LavanyaSrinivas06** [LavanyaSrinivas06/AI-First-Preauth-Fraud-Detection#61]
### Ticket ID: FPN-12
Phase: Phase 3 — Integration & Product Build
Branch Name: feature/FPN-12-api-model-integration

---

### Context
Connect trained models to the `/predict` endpoint and implement the hybrid decision engine described in the proposal:
`P_final = α * P_xgb + (1 - α) * P_ae`
with thresholding to return Approve/Review/Block decisions.

---

### Scope
- Load XGBoost model, Autoencoder, and AE threshold.
- Compute both model scores from the transformed payload.
- Implement weighted fusion (α default 0.7).
- Apply decision thresholds (`T_low`, `T_high`) and return final label.
- Return reason codes placeholder; SHAP-driven reasons will arrive in a later ticket.

---

### Steps

1. **Load artifacts**
   - In `api/service.py`:
     ```python
     import joblib, json
     from tensorflow.keras.models import load_model
     XGB = joblib.load("artifacts/xgb_model.pkl")
     AE = load_model("artifacts/autoencoder_model.h5")
     AE_T = float(Path("artifacts/ae_threshold.txt").read_text().strip())
     FEATURES = json.loads(Path("artifacts/features.json").read_text())
     ```
   - Keep α, `T_low`, `T_high` in `api/config.py` or environment variables:
     - `ALPHA = 0.7`
     - `T_LOW = 0.25`
     - `T_HIGH = 0.75`
     (adjust later based on evaluation)

2. **Build inference function**
   - Input: validated `TransactionIn` → pandas DataFrame with correct columns/order.
   - Transform using `PREPROCESS.transform`.
   - XGBoost:
     ```python
     p_xgb = float(XGB.predict_proba(X)[:,1][0])
     ```
   - Autoencoder anomaly score:
     ```python
     X_rec = AE.predict(X, verbose=0)
     err = float(((X - X_rec)**2).mean())
     # Map error to [0,1] anomaly score; simplest is percentile scaling or logistic mapping.
     p_ae = 1.0 if err >= AE_T else 0.0  # start with thresholded score
     ```
     Optionally implement a smooth mapping (e.g., min-max over validation errors) in a helper.

3. **Hybrid decision**
   - `p_final = ALPHA * p_xgb + (1 - ALPHA) * p_ae`
   - Decision:
     - if `p_final >= T_HIGH` → `block`
     - elif `p_final <= T_LOW` → `approve`
     - else → `review`

4. **Response schema**
   - Populate:
     ```json
     {
       "label": "...",
       "score_xgb": <float>,
       "score_ae": <float>,
       "ensemble_score": <float>,
       "reason_codes": []   // to be filled by SHAP ticket
     }
     ```

5. **Latency and logging**
   - Reuse latency header from FPN-11.
   - Log input hash (not raw PII), scores, and final decision to a rotating log file under `logs/inference.log`.

6. **Tests**
   - `tests/test_api_models.py`:
     - Mocks or uses small artifacts to test end-to-end scoring.
     - Asserts that boundary cases produce expected labels given fixed thresholds.
     - Ensures missing model files produce a controlled 500 with clear error.

7. **Docs**
   - Update `docs/api_usage.md` with:
     - Example request yielding each label (approve/review/block).
     - Explanation of hybrid scoring and thresholds.
     - Notes on configurable α, T_low, T_high.

---

### Acceptance Criteria
- [ ] `/predict` returns populated scores and final label from hybrid decision.
- [ ] Models and thresholds load once at startup; no per-request reload.
- [ ] Unit tests cover happy path and threshold boundary decisions.
- [ ] Inference logs record decisions without storing sensitive fields.
- [ ] Documentation updated with labeled examples and configuration notes.

---

### Artifacts
- `api/service.py` (model loading and inference)
- `api/config.py` (thresholds, alpha)
- `tests/test_api_models.py`
- `logs/inference.log` (runtime)
- Updated `docs/api_usage.md`

---

### Helpful Links
- FastAPI dependencies and startup events: https://fastapi.tiangolo.com/advanced/events/
- XGBoost predict_proba: https://xgboost.readthedocs.io/en/stable/python/python_api.html
- Keras model loading: https://keras.io/api/models/model_saving_apis/

---

### Effort Estimate
M (1–2 days)
